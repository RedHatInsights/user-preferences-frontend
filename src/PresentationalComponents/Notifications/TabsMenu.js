import React, { useMemo } from 'react';
import {
  Button,
  ButtonVariant,
  Divider,
  EmptyState,
  EmptyStateBody,
  EmptyStateFooter,
  EmptyStateHeader,
  EmptyStateIcon,
  EmptyStateVariant,
  Menu,
  MenuContent,
  MenuGroup,
  MenuItem,
  MenuList,
  MenuSearch,
  MenuSearchInput,
  SearchInput,
  TextInput,
} from '@patternfly/react-core';
import { SearchIcon } from '@patternfly/react-icons';
import { useLocation, useNavigate } from 'react-router-dom';
import PropTypes from 'prop-types';
import { getNavFromURL } from './urlSync';

const renderEmptyState = (setSearch) => (
  <EmptyState variant={EmptyStateVariant.sm} className="pf-u-mt-lg">
    <EmptyStateHeader
      titleText="No matching services found"
      icon={<EmptyStateIcon icon={SearchIcon} />}
      headingLevel="h4"
    />
    <EmptyStateBody>Adjust your filters and try again.</EmptyStateBody>
    <EmptyStateFooter>
      <Button variant={ButtonVariant.link} onClick={() => setSearch('')}>
        Clear filters
      </Button>
    </EmptyStateFooter>
  </EmptyState>
);

const TabsMenu = ({ searchRef, search, setSearch, fields, onClick }) => {
  const location = useLocation();
  const navigate = useNavigate();

  const { bundle, app } = useMemo(
    () => getNavFromURL(location, navigate, fields, {}),
    [location.search]
  );

  return (
    <Menu isPlain isScrollable>
      <MenuSearch>
        <MenuSearchInput ref={searchRef} className="pf-u-mx-sm">
          <SearchInput
            data-testid="search-input"
            aria-label="Filter menu items"
            placeholder="Search services"
            customIcon={<SearchIcon />}
            type="search"
            onChange={(_e, value) => setSearch(value)}
            onClear={(_e) => setSearch('')}
            value={search}
          />
        </MenuSearchInput>
      </MenuSearch>
      <Divider />
      <MenuContent id="notifications-menu-content">
        {fields.some((bundle) => bundle.fields.length > 0)
          ? fields.map(({ fields, title: bundleLabel, name: bundleName }) =>
              fields.length > 0 ? (
                <MenuGroup
                  label={bundleLabel}
                  className="pf-u-px-sm"
                  key={`menu-group-${bundleName}`}
                >
                  <MenuList>
                    {fields.map(
                      ({ label: sectionLabel, name: sectionName }) => (
                        <MenuItem
                          onClick={(e) => onClick(e, bundleName, sectionName)}
                          key={`menu-item-${bundleName}-${sectionName}`}
                          isFocused={
                            bundle === bundleName && app === sectionName
                          }
                        >
                          {sectionLabel}
                        </MenuItem>
                      )
                    )}
                  </MenuList>
                </MenuGroup>
              ) : null
            )
          : renderEmptyState(setSearch)}
      </MenuContent>
    </Menu>
  );
};

TabsMenu.propTypes = {
  fields: PropTypes.array.isRequired,
  search: PropTypes.string,
  setSearch: PropTypes.func,
  searchRef: PropTypes.object,
  onClick: PropTypes.func,
};

export default TabsMenu;
