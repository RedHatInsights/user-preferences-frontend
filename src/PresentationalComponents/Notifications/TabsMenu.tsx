import React, { useMemo } from 'react';
import {
  Button,
  ButtonVariant,
  Divider,
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStateVariant,
  Menu,
  MenuContent,
  MenuGroup,
  MenuList,
  TextInput,
  Title,
} from '@patternfly/react-core';
import { MenuItem, MenuSearch, MenuSearchInput } from '@patternfly/react-core/dist/dynamic/components/Menu';
import { SearchIcon } from '@patternfly/react-icons';
import { useLocation, useNavigate } from 'react-router-dom';
import PropTypes, { string } from 'prop-types';
import { getNavFromURL } from './urlSync';

const renderEmptyState = (setSearch: (searchString: string) => any) => (
  <EmptyState variant={EmptyStateVariant.sm} className="pf-u-mt-lg">
    <EmptyStateIcon icon={SearchIcon} />
    <Title headingLevel="h4" size="lg">
      No matching services found
    </Title>
    <EmptyStateBody>Adjust your filters and try again.</EmptyStateBody>
    <Button variant={ButtonVariant.link} onClick={() => setSearch('')}>
      Clear filters
    </Button>
  </EmptyState>
);

export type TabsMenuProps = {
  fields: any[],
  search: string,
  setSearch: (searchValue: any) => void,
  searchRef: any,
  onClick: (e: void, bundleName: string, sectionName: string) => void,
}

const TabsMenu = ({ searchRef, search, setSearch, fields, onClick }: TabsMenuProps) => {
  const location = useLocation();
  const navigate = useNavigate();

  const { bundle, app } = useMemo(
    () => getNavFromURL(location, navigate, fields, {}),
    [location.search]
  );

  return (
    <Menu isPlain isScrollable>
      <MenuSearch>
        <MenuSearchInput ref={searchRef} className="pf-u-mx-sm">
          <TextInput
            aria-label="Filter menu items"
            placeholder="Search services"
            customIcon="search"
            type="search"
            onChange={(value) => setSearch(value)}
            value={search}
          />
        </MenuSearchInput>
      </MenuSearch>
      <Divider />
      <MenuContent id="notifications-menu-content">
        {fields.some((bundle) => bundle.fields.length > 0)
          ? fields.map(({ fields, title: bundleLabel, name: bundleName }) =>
              fields.length > 0 ? (
                <MenuGroup
                  label={bundleLabel}
                  className="pf-u-px-sm"
                  key={`menu-group-${bundleName}`}
                >
                  <MenuList>
                    {fields.map(
                      ({ label: sectionLabel, name: sectionName }: {label: string, name: string}) => (
                        <MenuItem
                          onClick={(e) => onClick(e, bundleName, sectionName)}
                          key={`menu-item-${bundleName}-${sectionName}`}
                          isFocused={
                            bundle === bundleName && app === sectionName
                          }
                        >
                          {sectionLabel}
                        </MenuItem>
                      )
                    )}
                  </MenuList>
                </MenuGroup>
              ) : null
            )
          : renderEmptyState(setSearch)}
      </MenuContent>
    </Menu>
  );
};

export default TabsMenu;
