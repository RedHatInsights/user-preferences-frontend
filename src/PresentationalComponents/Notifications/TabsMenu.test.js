import React from 'react';
import { Form, RendererContext } from '@data-driven-forms/react-form-renderer';
import {
  fireEvent,
  getByRole,
  getByText,
  render,
} from '@testing-library/react';
import TabsMenu from './TabsMenu';

describe('TabsMenu tests', () => {
  const setSearch = jest.fn();
  const fields = [
    {
      title: 'Red Hat Enterprise Linux',
      name: 'rhel',
      fields: [
        {
          name: 'advisor',
          label: 'Advisor',
          component: 'tabGroup',
          fields: [],
          bundle: 'rhel',
        },
      ],
    },
  ];

  afterEach(() => {
    setSearch.mockReset();
  });

  it('should render empty state correctly', () => {
    const { container } = render(
      <Form onSubmit={() => undefined}>
        {() => (
          <RendererContext.Provider
            value={{
              formOptions: {
                renderForm: () => {},
              },
            }}
          >
            <TabsMenu search="" fields={[]} isLoading={false} />
          </RendererContext.Provider>
        )}
      </Form>
    );
    expect(container).toMatchSnapshot();
  });
  it('should clear search from the empty state on click', () => {
    const { container } = render(
      <Form onSubmit={() => undefined}>
        {() => (
          <RendererContext.Provider
            value={{
              formOptions: {
                renderForm: () => {},
              },
            }}
          >
            <TabsMenu
              search=""
              fields={[]}
              isLoading={false}
              setSearch={setSearch}
            />
          </RendererContext.Provider>
        )}
      </Form>
    );
    fireEvent.click(getByText(container, 'Clear filters'));
    expect(setSearch).toBeCalledTimes(1);
    expect(setSearch).toBeCalledWith('');
  });
  it('should render nav correctly', () => {
    const { container } = render(
      <Form onSubmit={() => undefined}>
        {() => (
          <RendererContext.Provider
            value={{
              formOptions: {
                renderForm: () => {},
              },
            }}
          >
            <TabsMenu
              search=""
              setSearch={() => null}
              fields={fields}
              isLoading={false}
              onClick={() => null}
            />
          </RendererContext.Provider>
        )}
      </Form>
    );
    expect(container).toMatchSnapshot();
  });
  it('should set search on input typing', () => {
    const { container } = render(
      <Form onSubmit={() => undefined}>
        {() => (
          <RendererContext.Provider
            value={{
              formOptions: {
                renderForm: () => {},
              },
            }}
          >
            <TabsMenu
              search=""
              setSearch={setSearch}
              fields={fields}
              isLoading={false}
              onClick={() => null}
            />
          </RendererContext.Provider>
        )}
      </Form>
    );
    fireEvent.change(getByRole(container, 'searchbox'), {
      target: { value: 'someText' },
    });
    expect(setSearch).toBeCalledTimes(1);
    expect(setSearch).toBeCalledWith('someText');
  });
  it('should call onClick callback correctly', () => {
    const onClick = jest.fn();
    const { container } = render(
      <Form onSubmit={() => undefined}>
        {() => (
          <RendererContext.Provider
            value={{
              formOptions: {
                renderForm: () => {},
              },
            }}
          >
            <TabsMenu
              search=""
              fields={fields}
              isLoading={false}
              onClick={onClick}
            />
          </RendererContext.Provider>
        )}
      </Form>
    );
    fireEvent.click(getByText(container, 'Advisor'));
    expect(onClick).toBeCalledTimes(1);
    expect(onClick).toBeCalledWith(expect.anything(), 'rhel', 'advisor');
  });
});
