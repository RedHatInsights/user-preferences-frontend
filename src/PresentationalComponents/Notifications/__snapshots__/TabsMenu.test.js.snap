// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabsMenu tests should render empty state correctly 1`] = `
<div>
  <div
    class="pf-v5-c-menu pf-m-plain pf-m-scrollable"
    data-ouia-component-id="OUIA-Generated-Menu-1"
    data-ouia-component-type="PF5/Menu"
    data-ouia-safe="true"
  >
    <div
      class="pf-v5-c-menu__search"
    >
      <div
        class="pf-v5-c-menu__search-input pf-u-mx-sm"
      >
        <span
          class="pf-v5-c-form-control pf-m-icon"
        >
          <input
            aria-invalid="false"
            aria-label="Filter menu items"
            data-ouia-component-id="OUIA-Generated-TextInputBase-1"
            data-ouia-component-type="PF5/TextInput"
            data-ouia-safe="true"
            placeholder="Search services"
            type="search"
            value=""
          />
          <span
            class="pf-v5-c-form-control__utilities"
          >
            <span
              class="pf-v5-c-form-control__icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                />
              </svg>
            </span>
          </span>
        </span>
      </div>
    </div>
    <hr
      class="pf-v5-c-divider"
    />
    <div
      class="pf-v5-c-menu__content"
      id="notifications-menu-content"
    >
      <div
        class="pf-v5-c-empty-state pf-m-sm pf-u-mt-lg"
      >
        <div
          class="pf-v5-c-empty-state__content"
        >
          <div
            class="pf-v5-c-empty-state__header"
          >
            <div
              class="pf-v5-c-empty-state__icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                />
              </svg>
            </div>
            <div
              class="pf-v5-c-empty-state__title"
            >
              <h4
                class="pf-v5-c-empty-state__title-text"
              >
                No matching services found
              </h4>
            </div>
          </div>
          <div
            class="pf-v5-c-empty-state__body"
          >
            Adjust your filters and try again.
          </div>
          <div
            class="pf-v5-c-empty-state__footer"
          >
            <button
              aria-disabled="false"
              class="pf-v5-c-button pf-m-link"
              data-ouia-component-id="OUIA-Generated-Button-link-1"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              Clear filters
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TabsMenu tests should render nav correctly 1`] = `
<div>
  <div
    class="pf-v5-c-menu pf-m-plain pf-m-scrollable"
    data-ouia-component-id="OUIA-Generated-Menu-3"
    data-ouia-component-type="PF5/Menu"
    data-ouia-safe="true"
  >
    <div
      class="pf-v5-c-menu__search"
    >
      <div
        class="pf-v5-c-menu__search-input pf-u-mx-sm"
      >
        <span
          class="pf-v5-c-form-control pf-m-icon"
        >
          <input
            aria-invalid="false"
            aria-label="Filter menu items"
            data-ouia-component-id="OUIA-Generated-TextInputBase-3"
            data-ouia-component-type="PF5/TextInput"
            data-ouia-safe="true"
            placeholder="Search services"
            type="search"
            value=""
          />
          <span
            class="pf-v5-c-form-control__utilities"
          >
            <span
              class="pf-v5-c-form-control__icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                />
              </svg>
            </span>
          </span>
        </span>
      </div>
    </div>
    <hr
      class="pf-v5-c-divider"
    />
    <div
      class="pf-v5-c-menu__content"
      id="notifications-menu-content"
    >
      <section
        class="pf-v5-c-menu__group pf-u-px-sm"
      >
        <h1
          class="pf-v5-c-menu__group-title"
        >
          Red Hat Enterprise Linux
        </h1>
        <ul
          class="pf-v5-c-menu__list"
          role="menu"
        >
          <li
            class="pf-v5-c-menu__list-item"
            role="none"
          >
            <button
              class="pf-v5-c-menu__item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              <span
                class="pf-v5-c-menu__item-main"
              >
                <span
                  class="pf-v5-c-menu__item-text"
                >
                  Advisor
                </span>
              </span>
            </button>
          </li>
        </ul>
      </section>
    </div>
  </div>
</div>
`;
